kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/ksp/TargetTypes.kt

```diff
@@ -0,0 +1,283 @@
+    visibility = resolvedVisibility
+  )
+}
+
+private fun ClassName.withTypeArguments(arguments: List<TypeName>): TypeName {
+  return if (arguments.isEmpty()) {
+    this
+  } else {
+    this.parameterizedBy(arguments)
+  }
+}
+
+@OptIn(KspExperimental::class)
+internal fun primaryConstructor(
+  resolver: Resolver,
+  targetType: KSClassDeclaration,
+  typeParameterResolver: TypeParameterResolver,
+  logger: KSPLogger
+): TargetConstructor? {
+  val primaryConstructor = targetType.primaryConstructor ?: return null
+
+  val parameters = LinkedHashMap<String, TargetParameter>()
+  for ((index, parameter) in primaryConstructor.parameters.withIndex()) {
+    val name = parameter.name!!.getShortName()
+    parameters[name] = TargetParameter(
+      name = name,
+      index = index,
+      type = parameter.type.toTypeName(typeParameterResolver),
+      hasDefault = parameter.hasDefault,
+      qualifiers = parameter.qualifiers(resolver),
+      jsonName = parameter.jsonName(resolver)
+    )
+  }
+
+  val kmConstructorSignature: String = resolver.mapToJvmSignature(primaryConstructor)
+    ?: run {
+      logger.error("No primary constructor found.", primaryConstructor)
+      return null
+    }
+  return TargetConstructor(
+    parameters, primaryConstructor.getVisibility().asKModifier(),
+    kmConstructorSignature
+  )
+}
+
+private fun KSAnnotated?.qualifiers(resolver: Resolver): Set<AnnotationSpec> {
+  if (this == null) return setOf()
+  val jsonQualifierType = resolver.getClassDeclarationByName<JsonQualifier>().asType()
+  return annotations
+    .filter {
+      it.annotationType.resolve().declaration.findAnnotationWithType(jsonQualifierType) != null
+    }
+    .mapTo(mutableSetOf()) {
+      it.toAnnotationSpec(resolver)
+    }
+}
+
+private fun KSAnnotated?.jsonName(resolver: Resolver): String? {
+  if (this == null) return null
+  val jsonType = resolver.getClassDeclarationByName<Json>().asType()
+  return findAnnotationWithType(jsonType)?.getMember<String>("name")
+}
+
+private fun declaredProperties(
+  constructor: TargetConstructor,
+  originalType: KSClassDeclaration,
+  classDecl: KSClassDeclaration,
+  resolver: Resolver,
+  typeParameterResolver: TypeParameterResolver,
+): Map<String, TargetProperty> {
+
+  val result = mutableMapOf<String, TargetProperty>()
+  for (property in classDecl.getDeclaredProperties()) {
+    val initialType = property.type.resolve()
+    val resolvedType = if (initialType.declaration is KSTypeParameter) {
+      property.asMemberOf(originalType.asType())
+    } else {
+      initialType
+    }
+    val propertySpec = property.toPropertySpec(resolver, resolvedType, typeParameterResolver)
+    val name = propertySpec.name
+    val parameter = constructor.parameters[name]
+    result[name] = TargetProperty(
+      propertySpec = propertySpec,
+      parameter = parameter,
+      visibility = property.modifiers.map { KModifier.valueOf(it.name) }.visibility(),
```

Comment:
Any reason not to use `property.getVisibility().asKModifier()` here?

Reply:
Ah good find! d808086